### DISCUSSION TITLE
Clarification of terminology in Etsy V3 API documentation

### DISCUSSION THREAD

**Discussion by alexandergx** [etsy/open-api#972] [Q&A]
Could someone from the Etsy team clarify some of the terminology used in the V3 API? It seems the language is quite unique to Etsy's systems.

For example, the API documentation uses terms like "listings", "offerings", and "property_values", whereas BigCommerce, Shopify, and other major platforms may use clearer and simpler terms like "products" and "variants". In this case, what would "offering" and "property_values" refer to? Are one of these meant to reference "product variants"?

The API documentation also references "receipts", whereas other major platforms may use a more standard term like "orders". And in accounting terms something like "line items" would be an important resource. How can we deduce a "line item" of an order from the "receipts" API?

**Comment by DrewHang** [comment]
Good question, I'm curious about this as well.

**Comment by kenxu0927** [comment]
I wonder if there's any API objects that correspond to "line items" as the other platforms commonly use

**Comment by seunggs** [comment]
Clarification on this would be helpful for me as well

**Comment by Gareth064** [comment]
| Etsy | Shopify |
| ------------- | ------------- |
| Receipt  | Order  |
| Transaction  | Line Item  |
| Listing | Product  |
| Product + Offering | Variation  |

I will add more to this when I get a chance.

But FYI - Product and Offering property_values, property_ID's etc and all that stuff.... very very complex in Etsy. If you look through this repo's discussion, you will find lots of posts on about how to use them. It will be a lot of trial and error.

**Reply by alexandergx** [reply]
This is really helpful. There's a couple things I'm still unclear about:

A "listing" is a "product", but if we look at the example response from `https://openapi.etsy.com/v3/application/shops/{shop_id}/receipts/{receipt_id}/transactions` we see it has both a "listing_id" and a "product_id". In the example it also suggests they're unique, or at least not the same, as they have different values. So it's not clear which represents the de-facto "product ID".

```
{
  "count": 0,
  "results": [
    {
      ...
      "listing_id": 0,
      ...
      "product_id": 1,
      ...
    }
  ]
}
```

**Reply by alexandergx** [reply]
Further, a "transaction" / "line item" of a "receipt" / "order" has the returned properties `variations[]` and `product_data[]`:

```
{
  "count": 0,
  "results": [
  ...
    {
      "variations": [
        {
          "property_id": 0,
          "value_id": 0,
          "formatted_name": "string",
          "formatted_value": "string"
        }
      ],
      "product_data": [
        {
          "property_id": 1,
          "property_name": "string",
          "scale_id": 1,
          "scale_name": "string",
          "value_ids": [
            1
          ],
          "values": [
            "string"
          ]
        }
      ],
      ...
    }
  ]
}
```

As your legend states, "variation" = "offering" + "property_values"? If we look at a response from `https://openapi.etsy.com/v3/application/listings/{listing_id}/inventory/products/{product_id}` we get:

```
{
  ...
  "offerings": [
    {
      "offering_id": 1,
      "quantity": 0,
      "is_enabled": true,
      "is_deleted": true,
      "price": {
        "amount": 0,
        "divisor": 0,
        "currency_code": "string"
      }
    }
  ],
  "property_values": [
    {
      "property_id": 1,
      "property_name": "string",
      "scale_id": 1,
      "scale_name": "string",
      "value_ids": [
        1
      ],
      "values": [
        "string"
      ]
    }
  ]
}
```

`property_values` data corresponds with `product_data` but they have different naming conventions. And `variations` data does not correspond with `offerings` data, but it's naming convention suggests that it should correspond to product variations.

**Reply by Gareth064** [reply]
Random question, but are you trying to figure out Etsy's API just by looking at their Documentation? If so, then you will struggle to understand it. You're better off building something basic and connect it to a shop, start making calls to see what the actual data looks like. When you see real data, you can match it up to what you see in the GUI.

I know documentation should provide you all the info you need, but unfortunately the Etsy one is very complex and requires that you can get access to a shops data to understand it fully.

**Reply by alexandergx** [reply]
Exactly. Although, I should say the documentation has been accurate so far as to what's returned by the API. My main questions are pertaining to the legibility of the data, rather than the integrity of the documentation.

**Reply by emergingdzns** [reply]
> This is really helpful. There's a couple things I'm still unclear about:
> 
> A "listing" is a "product", but if we look at the example response from `https://openapi.etsy.com/v3/application/shops/{shop_id}/receipts/{receipt_id}/transactions` we see it has both a "listing_id" and a "product_id". In the example it also suggests they're unique, or at least not the same, as they have different values. So it's not clear which represents the de-facto "product ID".
> ```
Sort of... In Etsy a listing is really a class or group of products so to speak. The listing has the ability to have many products in it. Each product is defined by the matrix of configurations in the system. So let's say you are selling a homemade mug. Each mug can come in 3 different sizes and 3 different colors. That means that what you have is a listing composed of 9 possible products. I look at products as variants. So listing_id 1 could have 9 product_ids 1-9. The offering is the meta data of the product: price, sku, etc. The property_values define what makes that product unique (in the example above - size and color).

> Further, a "transaction" / "line item" of a "receipt" / "order" has the returned properties variations[] and product_data[]

Yes. So in the receipt (aka order) the line items are the transactions as mentioned before. The product data and variation data is returned as part of that transaction to ensure that the correct specific product is connected and not just the listing id, which is too vague if there are variants.

The naming of these different variables harkens back to the earlier days of Etsy I am sure when things were far more simple and there wasn't really a sort of "standard" in the industry. When the API was built, the naming followed otherwise it would confuse future internal developers because the internal system would have different names from the external API which would make for a larger learning curve for employees.
> This is really helpful. There's a couple things I'm still unclear about:
> 
> A "listing" is a "product", but if we look at the example response from `https://openapi.etsy.com/v3/application/shops/{shop_id}/receipts/{receipt_id}/transactions` we see it has both a "listing_id" and a "product_id". In the example it also suggests they're unique, or at least not the same, as they have different values. So it's not clear which represents the de-facto "product ID".
> ```
Sort of... In Etsy a listing is really a class or group of products so to speak. The listing has the ability to have many products in it. Each product is defined by the matrix of configurations in the system. So let's say you are selling a homemade mug. Each mug can come in 3 different sizes and 3 different colors. That means that what you have is a listing composed of 9 possible products. I look at products as variants. So listing_id 1 could have 9 product_ids 1-9. The offering is the meta data of the product: price, sku, etc. The property_values define what makes that product unique (in the example above - size and color).

> Further, a "transaction" / "line item" of a "receipt" / "order" has the returned properties variations[] and product_data[]

Yes. So in the receipt (aka order) the line items are the transactions as mentioned before. The product data and variation data is returned as part of that transaction to ensure that the correct specific product is connected and not just the listing id, which is too vague if there are variants.

The naming of these different variables harkens back to the earlier days of Etsy I am sure when things were far more simple and there wasn't really a sort of "standard" in the industry. When the API was built, the naming followed otherwise it would confuse future internal developers because the internal system would have different names from the external API which would make for a larger learning curve for employees.

**Reply by Gareth064** [reply]
This golden information and the type of explanation the documentation could benefit from. You don't want to go back and work there again @emergingdzns for a few months to write this stuff up? 🤣

**Comment by emergingdzns** [comment]
I’d love to go back to Etsy! So many things I didn’t have time for before my contract was up.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
